pkg/aws/awscluster/mutate_awscluster.go

```diff
@@ -0,0 +1,95 @@
+// Package awsmachinedeployment intercepts write activity to AWSMachineDeployment objects.
+package awscluster
+
+import (
+	"fmt"
+
+	infrastructurev1alpha2 "github.com/giantswarm/apiextensions/v2/pkg/apis/infrastructure/v1alpha2"
+	"github.com/giantswarm/k8sclient/v4/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	admissionv1 "k8s.io/api/admission/v1"
+
+	"github.com/giantswarm/aws-admission-controller/config"
+	"github.com/giantswarm/aws-admission-controller/pkg/mutator"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+// Mutator for AWSMachineDeployment object.
+type Mutator struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+
+	podCIDRBlock string
+}
+
+func NewMutator(config config.Config) (*Mutator, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	mutator := &Mutator{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		podCIDRBlock: fmt.Sprintf("%s/%s", config.PodSubnet, config.PodCIDR),
+	}
+
+	return mutator, nil
+}
+
+// Mutate is the function executed for every matching webhook request.
+func (m *Mutator) Mutate(request *admissionv1.AdmissionRequest) ([]mutator.PatchOperation, error) {
+	var result []mutator.PatchOperation
+
+	if request.DryRun != nil && *request.DryRun {
+		return result, nil
+	}
+	awsCluster := &infrastructurev1alpha2.AWSCluster{}
+	if _, _, err := mutator.Deserializer.Decode(request.Object.Raw, nil, awsCluster); err != nil {
+		return nil, microerror.Maskf(parsingFailedError, "unable to parse AWSCluster: %v", err)
+	}
+	if &awsCluster.Spec.Provider.Pods != nil {
+		if awsCluster.Spec.Provider.Pods.CIDRBlock != "" {
+			return result, nil
+		}
+		if awsCluster.Spec.Provider.Pods.ExternalSNAT != nil {
+			// If the Pod CIDR is not set but the pods attribute exists, we default here
+			m.Log("level", "debug", "message", fmt.Sprintf("AWSCluster %s Pod CIDR Block is not set and will be defaulted to %s",
+				awsCluster.ObjectMeta.Name,
+				m.podCIDRBlock),
+			)
+			patch := mutator.PatchAdd("/spec/provider/pods/", "cidrBlock")
+			result = append(result, patch)
+			patch = mutator.PatchAdd("/spec/provider/pods/cidrBlock", m.podCIDRBlock)
+			result = append(result, patch)
+			return result, nil
+		}
+	}
+	// If the Pod CIDR is not set we default it here
+	m.Log("level", "debug", "message", fmt.Sprintf("AWSCluster %s Pod CIDR Block is not set and will be defaulted to %s",
+		awsCluster.ObjectMeta.Name,
+		m.podCIDRBlock),
+	)
+	patch := mutator.PatchAdd("/spec/provider/", "pods")
+	result = append(result, patch)
+	patch = mutator.PatchAdd("/spec/provider/pods", map[string]string{"cidrBlock": m.podCIDRBlock})
```

Comment:
Did you try to use one mutator.PatchAdd with map[string]map[string]string{"pods": {cidrBlock: m.podCIDRBlock}} ?

Reply:
I don't think I have tried it but it might work. Do you think it's better to have just one patch?

Reply:
I was just curious if we could have one instead of having two, just because of simplicity and potential unexpected behaviours. If it works it's all good, you can still test this after merging no blocker from my side 🙂.

Reply:
Let's give it a try when we add more logic for https://github.com/giantswarm/roadmap/issues/198